---
title: "Priego-Cubero et al. Figure 1A"
output: html_document
---

# Load libraries 
```{r setup, include=FALSE}
library(tidyverse)
library(gggenes)
library(patchwork)
library(khroma)
```



# (Figure 1A) GENE CLUSTERS 

## Reformat annotations

Download all the annotations below and place them in the `input` folder 

#### O. sativa 

Get annotation from:

```
https://data.jgi.doe.gov/refine-download/phytozome?genome_id=323&_gl=1*d6rwwo*_ga*Nzc0NDYxMTE4LjE2OTUzODAyNTk.*_ga_YBLMHYR3C2*MTcwMjI5OTkwNi4zLjAuMTcwMjI5OTkwNi4wLjAuMA..&expanded=Phytozome-323
```

```{r}
O_sativa <-
  read_table('input/Osativa_323_v7.0.gene_exons.gff3', 
           skip = 3, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "Chr4") %>%
  filter(Start %in% 5318060:5490000) %>%
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, Gene) %>%
  mutate(Gene = gsub("Name=", "", Gene)) %>% 
  mutate(Genome = "O_sativa") %>%
  mutate(Gene_Name = case_when(
    Gene == "LOC_Os04g09900" ~ "CPS4",
    Gene %in% c("LOC_Os04g09920", "LOC_Os04g10160") ~ "CYP99A",
    Gene %in% c("LOC_Os04g10000", "LOC_Os04g10010") ~ "MS",
    Gene == "LOC_Os04g10060" ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%

# To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
```

#### O. punctata

Get annotation for chromosome 4 from: 
```
https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-57/gff3/oryza_punctata/
```

```{r}
O_punctata <-
  read_table('input/Oryza_punctata.Oryza_punctata_v1.2.56.chromosome.4.gff3', 
           skip = 7, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "4") %>%
  filter(Start %in% 4522665:4692353) %>%
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=gene:", "", ID)) %>% 
  mutate(Genome = "O_punctata") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  mutate(Gene_Name = case_when(
    Gene == "OPUNC04G02770" ~ "CPS4",
    Gene %in% c("OPUNC04G02780", "OPUNC04G02870") ~ "CYP99A",
    Gene %in% c("OPUNC04G02800", "OPUNC04G02820") ~ "MS",
    Gene == "OPUNC04G02860" ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%
# To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
  
```

#### O. officinalis

Get annotation from: 

```
https://datacommons.cyverse.org/browse/iplant/home/shared/commons_repo/curated/Shenton_OofficinalisAnnotation_2019
```

```{r}
O_officinalis <-
  read_table('input/Ooffi_maker_gene_annotation.gff', 
           #skip = 7, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "OoffiChr04") %>%
  filter(Start %in% 7667067:8034073) %>% # beginning of the first and last gene in the cluster
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_officinalis") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  mutate(Gene_Name = case_when(
    Gene == "OoffiChr04g00263" ~ "CPS4",
    Gene %in% c("OoffiChr04g00260", "OoffiChr04g00264") ~ "CYP99A",
    Gene %in% c("OoffiChr04g00266") ~ "MS",
    Gene %in% c("OoffiChr04g00265", "OoffiChr04g00269") ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%
  # To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
```


#### O. eichingeri 

Get the annotation from: 

```
https://datacommons.cyverse.org/browse/iplant/home/shared/commons_repo/curated/Shenton_OofficinalisAnnotation_2019
```

Genes 
CPS4 maker-scaffold957-snap-gene-0.3-mRNA-1
CYP99A maker-scaffold6560-augustus-gene-0.2-mRNA-1 maker-scaffold300-snap-gene-0.9-mRNA-1
MS augustus_masked-scaffold6635-processed-gene-0.0-mRNA-1 maker-scaffold7787-augustus-gene-0.3-mRNA-1
KSL4 maker-scaffold7214-snap-gene-0.0-mRNA-1 maker-scaffold5747-snap-gene-0.4-mRNA-1

```{r}
O_eichingeri <-
  read_table('input/Oeich_maker_gene_annotation.gff', 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr %in% c("scaffold957", "scaffold6560", "scaffold300", "scaffold6635", "scaffold7787", "scaffold7214", "scaffold5747")) %>%
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_eichingeri") %>%
  select(!ID) %>%
  mutate(Gene_Name = case_when(
    Gene == "maker-scaffold957-snap-gene-0.3" ~ "CPS4",
    Gene %in% c("maker-scaffold6560-augustus-gene-0.2",
                "maker-scaffold300-snap-gene-0.9") ~ "CYP99A",
    Gene %in% c("augustus_masked-scaffold6635-processed-gene-0.0",
                "maker-scaffold7787-augustus-gene-0.3") ~ "MS",
    Gene %in% c("maker-scaffold7214-snap-gene-0.0", 
                "maker-scaffold5747-snap-gene-0.4") ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%
  # Filter anything that is not a momilactone gene  (to simplify the plot)
  filter(Gene_Name != "Intervening") %>%
  # I will format the position of the genes so that they appear one after the other
  mutate(length = abs(Start - End)) %>%
  mutate(Start = 1) %>%
  arrange(length)

# To prevent overlapping of genes from different scaffolds with similar start and end positions

  for (i in 2:nrow(O_eichingeri)){
    # Start (gene i) = Start (gene i-1) + Length(gene i-1) + 200 (just space in between)
    O_eichingeri[i, 2] = O_eichingeri[i-1, 2] + O_eichingeri[i-1, 8] + 30000
    # End (gene i) = Start (gene i) + Length(gene i)
    O_eichingeri[i, 3] = O_eichingeri[i, 2] + O_eichingeri[i, 8]
  }

  O_eichingeri[1,3] = O_eichingeri[1,2] + O_eichingeri[1,8]

  O_eichingeri <- O_eichingeri %>%
  select(!length)
```

#### O. rhizomatis

Get annotation from: 

```
https://datacommons.cyverse.org/browse/iplant/home/shared/commons_repo/curated/Shenton_OofficinalisAnnotation_2019
```

CPS4: maker-scaffold1491-snap-gene-0.13-mRNA-1
KSL4: maker-scaffold1491-snap-gene-0.11-mRNA-1 maker-scaffold2280-snap-gene-0.5-mRNA-1
MS: maker-scaffold3149-snap-gene-0.3-mRNA-1
CYP99A: maker-scaffold2280-augustus-gene-0.6-mRNA-1

```{r}
O_rhizomatis <-
  read_table('input/Orhiz_maker_gene_annotation.gff', 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr %in% c("scaffold1491", "scaffold3149", "scaffold2280")) %>%
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_rhizomatis") %>%
  select(!ID) %>%
  mutate(Gene_Name = case_when(
    Gene == "maker-scaffold1491-snap-gene-0.13" ~ "CPS4",
    Gene %in% c("maker-scaffold2280-augustus-gene-0.6") ~ "CYP99A",
    Gene %in% c("maker-scaffold3149-snap-gene-0.3") ~ "MS",
    Gene %in% c("maker-scaffold1491-snap-gene-0.11", 
                "maker-scaffold2280-snap-gene-0.5") ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%
  # Filter anything that is not a momilactone gene  (to simplify the plot)
  filter(Gene_Name != "Intervening") %>%
  # I will format the position of the genes so that they appear one after the other
  mutate(length = abs(Start - End)) %>%
  mutate(Start = 1) %>%
  arrange(length)

# To prevent overlapping of genes from different scaffolds with similar start and end positions: 

  for (i in 2:nrow(O_rhizomatis)){
    # Start (gene i) = Start (gene i-1) + Length(gene i-1) + 200 (just space in between)
    O_rhizomatis[i, 2] = O_rhizomatis[i-1, 2] + O_rhizomatis[i-1, 8] + 30000
    # End (gene i) = Start (gene i) + Length(gene i)
    O_rhizomatis[i, 3] = O_rhizomatis[i, 2] + O_rhizomatis[i, 8]
  }

  O_rhizomatis[1,3] = O_rhizomatis[1,2] + O_rhizomatis[1,8]

  O_rhizomatis <- O_rhizomatis %>%
  select(!length)
```

#### O. alta

Annotation from: 
```
https://download.cncb.ac.cn/gwh/Plants/Oryza_PPR1_GWHAZTO00000000/GWHAZTO00000000.gff.gz
```
Protein_Accession and (ID)

CPS4: GWHPAZTO032323 (OalC04g148650), GWHPAZTO032314 (OalC04g148770), GWHPAZTO032307 (OalC04g148840)
KSL4: GWHPAZTO032312 (OalC04g148800)
CYP99A: GWHPAZTO032327 (OalC04g148580)
MS: GWHPAZTO032310 (OalC04g148820)


```{r}
O_alta <-
  read_table('input/O_alta_GWHAZTO00000000.gff', 
           skip = 1, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "GWHAZTO00000004") %>%
  filter(Start %in% 4117988:4644751) %>% # beginning of the first and last gene in the cluster
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_alta") %>%
  type_convert() %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  mutate(Gene_Name = case_when(
    Gene %in% c("OalC04g148650", "OalC04g148720", "OalC04g148840") ~ "CPS4",
    Gene %in% c("OalC04g148580") ~ "CYP99A",
    Gene %in% c("OalC04g148820") ~ "MS",
    Gene %in% c("OalC04g148800") ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%  
  # To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         ) %>%
  # I will reduce the size of this cluster from 400 kb to match the 180 kb size of the others, improving overall clarity of the plot
  mutate(Start = case_when(
    Start %in% 350456:526764 ~ Start - 150000,
    Gene %in% c("OalC04g148770", "OalC04g148780", "OalC04g148750") ~ Start - 80000,
    TRUE ~ Start )
  ) %>% 
  mutate(End = case_when(
    End %in% 351589:528720 ~ End - 150000,
    Gene %in% c("OalC04g148770", "OalC04g148780", "OalC04g148750") ~ End - 80000,
    TRUE ~ End )
  )
```
#### O. coarctata LL

CPS4: Oco08G001460.1
KSL4: Oco08G001500.1
MS: Oco08G001470.1
CYP99A: Oco08G001510.1
CYP76L11: Oco08G001480.1

```{r}
O_coarctata_L <-
  read_table('/dss/dsslegfs01/pn73so/pn73so-dss-0000/lab/santi/genomes/O_coarctata/figshare/Oco_Chr_genome_all.gff', 
           skip = 19, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "LG08") %>%
  filter(Start %in% 2283517:2338065) %>%
  select(Chr, Start, End, Orientation, Metadata) %>%
  mutate(Gene = gsub("ID=", "", Metadata)) %>%
  select(!Metadata) %>%
  mutate(Genome = "OcoL") %>%
  type_convert() %>%
  mutate(Gene_Name = case_when(
    Gene == "Oco08G001460" ~ "CPS4",
    Gene %in% c("Oco08G001510") ~ "CYP99A",
    Gene %in% c("Oco08G001480") ~ "CYP76L11",
    Gene %in% c("Oco08G001470") ~ "MS",
    Gene == "Oco08G001500" ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>% 
# To have all the clusters on scale 
  
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
```


#### O. brachyantha

Annotation from 
```
https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/231/095/GCF_000231095.2_ObraRS2/GCF_000231095.2_ObraRS2_genomic.gff.gz
```

XP_040379471.1 = gene-LOC102721444
XP_006653173 = gene-LOC102721730
```{r}
O_brachyantha <-
  read_table('input/GCF_000231095.2_ObraRS2_genomic.gff', 
           skip = 8, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "NC_023166.2") %>%
  filter(Start %in% 2330641:2359703) %>%
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=gene-", "", ID)) %>% 
  mutate(Genome = "O_brachyantha") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  mutate(Gene_Name = case_when(
    Gene %in% c("LOC102721444", "LOC102721730") ~ "CYP99A",
    TRUE ~ "Intervening" )
  ) %>%
# To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
```


## Dummies for genomes without cluster

Just to represent assemblies with no MBGC detected

#### O. alta D

```{r}
O_alta_D <- data_frame(Chr = 'Chr4',
                              Start = 20000,
                              End = 50000,
                              Orientation = '+',
                              Gene = 'None',
                              Genome = 'OaltaD',
                              Gene_Name = 'Intervening')
```

#### O. australiensis

```{r}
O_australiensis <- data_frame(Chr = 'Chr4',
                              Start = 20000,
                              End = 50000,
                              Orientation = '+',
                              Gene = 'None',
                              Genome = 'O_australiensis',
                              Gene_Name = 'Intervening')
```

#### O_coarctata_K

```{r}
O_coarctata_K <- data_frame(Chr = 'Chr4',
                              Start = 20000,
                              End = 50000,
                              Orientation = '+',
                              Gene = 'None',
                              Genome = 'OcoK',
                              Gene_Name = 'Intervening')
```


#### O. granulata

```{r}
O_granulata <- data_frame(Chr = 'Chr4',
                              Start = 20000,
                              End = 50000,
                              Orientation = '+',
                              Gene = 'None',
                              Genome = 'O_granulata',
                              Gene_Name = 'Intervening')
```

#### L. perrieri

```{r}
L_perrieri <- data_frame(Chr = 'Chr4',
                              Start = 20000,
                              End = 50000,
                              Orientation = '+',
                              Gene = 'None',
                              Genome = 'L_perrieri',
                              Gene_Name = 'Intervening')

```



## Genes plot

Palette used: 

```{r}
khroma::color('light')(8)
```

'#DDDDDD' color for intervening genes

```{r}
all_genomes <- rbind(O_sativa, O_punctata, O_officinalis, O_eichingeri, O_rhizomatis, O_alta_C, O_alta_D, O_australiensis, O_coarctata_K, O_coarctata_L, O_brachyantha, O_granulata, L_perrieri)

# Create a factor variable with desired facet order
all_genomes$Genome <- factor(all_genomes$Genome, levels = c("Osativa", "Opunctata", "Ooffi", "Orhizo", "Oeich", "OaltaC", "OaltaD", "Oaus", "OcoK", "OcoL", "Obra", "Ogra", "Lperrieri"))

clusters <- 

  all_genomes %>%
  # Right format for orientation+
  
  mutate(Orientation = str_replace(Orientation, "\\+", 'TRUE')) %>%
  mutate(Orientation = str_replace(Orientation, "\\-", 'FALSE')) %>%
  
  # I will increase the size of MS and CYP genes otherwise the barely are seen in the plot 
  
  mutate(End = case_when(
    Gene_Name %in% c("MS", "CYP99A", "CYP76L11") ~ End + 1000,
    TRUE ~ End ) ) %>%
  
  # Plot
  
  ggplot(
    aes(xmin = Start, 
        xmax = End, 
        y = Genome, 
        fill = Gene_Name, 
        forward = Orientation,
        label = Gene_Name)
  ) +
  geom_gene_arrow(
                  arrowhead_height = unit(5, "mm"), 
                  arrowhead_width = unit(1, "mm"),
                  arrow_body_height = grid::unit(5, "mm"),
                  linetype = 'solid',
                  size = 0.2) +
  facet_wrap(~Genome, 
             scales = "free_y", 
             ncol = 1,
             labeller = label_parsed) +
  scale_fill_manual(values = c("#EE8866","#BBCC33", "#44BB99", "#EBEBEB", "#EEDD88", "#99DDFF")) +
  theme_genes() +
  #theme(legend.position="none") +
  theme(panel.grid.major.y = element_line(colour = NULL)) + # , linetype = "dotted")) +
  theme(
    legend.text = element_text(size = 10),
    legend.title = element_blank(),
    legend.position = "bottom",
    axis.title.x=element_blank(),
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(), 
    axis.line.x = element_blank(),
    axis.title.y=element_blank(),
    #axis.text.y=element_blank(),
    axis.ticks.y=element_blank(), 
    axis.line.y = element_blank() )

clusters

```


In this chunck of code I will add white rectangles to give the plot the look that I want.
It is repetitive and not so elegant but I can't find a better solution to get exactly the figure I want.

```{r}
clusters <- clusters + # Add a white rectangle
  
# Add scaffold look for O. rhizomatis and O. eich
  
geom_rect(xmin = 10000, xmax = 20000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Orhizo", "Oeich")
)) +
  geom_rect(xmin = 40000, xmax = 50000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Oeich")
)) +
    geom_rect(xmin = 70000, xmax = 80000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Orhizo", "Oeich")
)) +
    geom_rect(xmin = 110000, xmax = 120000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Oeich")
)) +
      geom_rect(xmin = 140000, xmax = 150000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Oeich")
)) +
      geom_rect(xmin = 175000, xmax = 185000, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Oeich")
)) +
  
# Fragment line in O. alta to indicate that the region is too large
  
geom_rect(xmin = 120498, xmax = 185603, ymin = -2, ymax = 2, fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("OaltaC")
)) +


# Beautify genomes without cluster
  
geom_rect(
  xmin = 2000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("OaltaD", "Oaus", "OcoK", "Ogra", "Lperrieri")
)) +
  
# Remove extra 'grey' line to sativa, punctata, officinalis, rhizomatis and eich 

geom_rect(
  xmin = 200000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Osativa", "Opunctata")
)) +
geom_rect(
  xmin = 150000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Orhizo")
)) +
geom_rect(
  xmin = 220000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Oeich")
)) +
geom_rect(
  xmin = 90000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("OcoL")
)) +
geom_rect(
  xmin = 50000, xmax = 1000000,
  ymin = -2, ymax = 2,
  fill = "white", color = "white",
  data = ~.x %>% 
    filter(Genome %in% c("Obra")
))


clusters
```

Saving plot 
```{r}
ggsave('plots/clusters.pdf',
      plot = clusters,
      dpi = 400,
      width = 10, 
      height = 5) 
```

