---
title: "Figure 4E"
output: html_document
---

# Load libraries 
```{r setup, include=FALSE}
library(tidyverse)
library(gggenes)
library(khroma)
library(ggrepel)
library(MoMAColors)
```

# (Figure 4E) SNPs PLOT 

### Load SNP calling data

This code is for loading multiple files. However, in this particular case, it will only load one file
```{r}
snp_files <- list.files(path = "../oryza-phylogenetic-trees/input", 
                    pattern = ".tsv", 
                    full.names = T) %>% 
  as_tibble() %>%
  mutate(path = value) %>%
  separate(col = value, sep = "/", into = c('c1', 'c2', 'c3', 'c4')) %>%
  separate(col = c4, sep = ".tsv", into = c('snp_file')) %>%
  select(path, snp_file) %>%
  mutate(name_file = case_when(
    snp_file %in% c("officinalis.KSL4-2.merged.snps.snpeff.haplotypecaller") ~ "OoKSL4_2_HC",
    TRUE ~ snp_file)
  )

# Reading each file at once

for (file in 1:nrow(snp_files)) {
  assign(snp_files$name_file[file], 
         read_tsv(snp_files$path[file],
                  col_names = c('CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO'))
         )# The result of the function call ape::read.tree is assigned to the name using the assign() function
 
}
```

### Load O. officinalis and KSL4-2 annotation

Get annotation from: 

```
https://datacommons.cyverse.org/browse/iplant/home/shared/commons_repo/curated/Shenton_OofficinalisAnnotation_2019
```

```{r}
O_officinalis <-
  read_table('input/Ooffi_maker_gene_annotation.gff', 
           #skip = 7, 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Feature == "gene") %>%
  filter(Chr == "OoffiChr04") %>%
  filter(Start %in% 7667067:8034073) %>% # beginning of the first and last gene in the cluster
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_officinalis") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  mutate(Gene_Name = case_when(
    Gene == "OoffiChr04g00263" ~ "CPS4",
    Gene %in% c("OoffiChr04g00260", "OoffiChr04g00264") ~ "CYP99A",
    Gene %in% c("OoffiChr04g00266") ~ "MS",
    Gene %in% c("OoffiChr04g00265", "OoffiChr04g00269") ~ "KSL4",
    TRUE ~ "Intervening" )
  ) %>%
  # To have the clusters on scale 
  mutate(End = End - min(Start) + 1,
         Start = Start - min(Start) + 1
         )
```


```{r}
OoKSL4_2_gene <-
  read_table('/dss/dsslegfs01/pn73so/pn73so-dss-0000/lab/santi/genomes/annotation/Ooffi_maker_gene_annotation.gff', 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Chr == "OoffiChr04") %>%
  filter(Start %in% 7877099:7889563) %>% # beginning of the first and last gene in the cluster
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_officinalis") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  filter(Gene %in% c("OoffiChr04g00265")) %>%
  mutate(Gene_Name = case_when(
    Gene == "OoffiChr04g00265" ~ "OoKSL4-2",
    TRUE ~ "" ) )
    
# Check how the dataframe should look like to represent the different exons
# gggenes::example_subgenes    

OoKSL4_2_subgene <-
  read_table('input/Ooffi_maker_gene_annotation.gff', 
           col_names = c('Chr', 'Annotator', 'Feature', 'Start', 'End', 'c1', 'Orientation', 'c2', 'Metadata')) %>%
  filter(Chr == "OoffiChr04") %>%
  filter(Start %in% 7877099:7889563) %>% # beginning of the first and last gene in the cluster
  separate(Metadata, into = c("ID", "Gene"), sep = ";") %>%
  select(Chr, Start, End, Orientation, ID) %>%
  mutate(Gene = gsub("ID=", "", ID)) %>% 
  mutate(Genome = "O_officinalis") %>%
  select(!ID) %>%
  mutate(Chr = "Chr4") %>%
  separate(Gene, into = c("Gene", "Gene_Feature"), sep = ":") %>%
  filter(Gene %in% c("OoffiChr04g00265.1")) %>%
  filter(Gene_Feature == 'cds') %>%
  mutate(from = Start) %>%
  mutate(to = End) %>%
  mutate(Start = 7878906) %>%
  mutate(End = 7885876) %>%
  mutate(Gene_Name = case_when(
    Gene == "OoffiChr04g00265.1" ~ "OoKSL4-2",
    TRUE ~ "" ) )
```

## Plot KSL4-2
```{r}
KSL4_2_plot <- OoKSL4_2_gene %>%
  ggplot(
    aes(xmin = Start, 
        xmax = End, 
        y = Gene_Name)
  ) +
  geom_gene_arrow(
                  arrowhead_height = unit(10, "mm"), 
                  arrowhead_width = unit(1, "mm"),
                  arrow_body_height = grid::unit(10, "mm"),
                  linetype = 'solid',
                  size = 0.4,
                  color = "black",
                  fill = "white", 
                  alpha = .2)  +
  geom_subgene_arrow(data = OoKSL4_2_subgene,
                     aes(xmin = Start, 
                         xmax = End, 
                         y = Gene_Name, 
                         fill = Gene_Feature,
                         xsubmin = from, 
                         xsubmax = to),
                     fill = "#EEDD88",
                     color = "black", 
                     arrowhead_height = unit(10, "mm"), 
                     arrowhead_width = unit(1, "mm"),
                     arrow_body_height = grid::unit(10, "mm"),) +
  xlim(7878906, 7885876) +
  theme_genes() +
  theme(
    legend.text = element_text(size = 10),
    legend.title = element_blank(),
    legend.position = "bottom",
    axis.title.y=element_blank(),
    axis.ticks.y=element_blank(), 
    axis.line.y = element_blank(),
    axis.title.x=element_blank(),
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(), 
    axis.line.x = element_blank()
    ) +
  # Add SNP position
  geom_vline(xintercept = 7882960, color = "red", linetype = "dashed", linewidth = .5)

KSL4_2_plot

```

## Plot SNPs

AC = Allele count in genotypes for each ALT allele

For this same site, AC = 2 means that out of 15 samples (30 haplotypes), we saw the variant 2 times. Out of 30 haplotypes, this is a frequency of 0.067


```{r}
OoKSL4_2_HC_filtered <- OoKSL4_2_HC %>%
  # Filter out upstream_gene_variant and downstream_gene_variant
  filter(POS %in% 7879099:7885563) %>% 
  # Get first occurrence of modifier, low, high or moderate from INFO column
  mutate(Effect_impact = str_extract(INFO, "MODIFIER|LOW|HIGH|MODERATE")) %>%
  # Get the info preceeding the first occurrence of modifier, low, high or moderate from info
  mutate(Effect = str_extract(INFO, "(?<=\\|)[^|]+(?=\\|(MODIFIER|LOW|HIGH|MODERATE))")) %>%
  mutate(AC = str_extract(INFO, ";AC=\\d+")) %>%
  select(!INFO) %>%
  mutate(AC = gsub(";AC=", "", AC)) %>%
  mutate(Allele_Frequency = (as.numeric(AC)/2) / 15)
```


```{r}
SNPs_plot <- OoKSL4_2_HC_filtered %>%
  filter(!(Effect == 'intron_variant')) %>% 
  filter(Allele_Frequency > 0.2 | Effect_impact == 'HIGH') %>% # at least 3 / 15 have the SNP or have HIGH effect
  # I got the accessions by checking the .vcf file on IGV
  mutate(Accessions = case_when(
    POS == 7880557 ~ "W1291",
    POS == 7880559 ~ "W1291",
    POS == 7879329 ~ "W1813",
    POS == 7884110 ~ "W0065",
    POS == 7880299 ~ "W0065,W1302,W1200,W1315,W1308",
    POS == 7884201 ~ "W0065",
    POS == 7884301 ~ "W0065",
    TRUE ~ "" ) ) %>%
  ggplot(aes(x=POS, 
             y=Allele_Frequency,
             fill = Effect,
             label = Accessions)) + 
  geom_point(pch = 21,
             size = 1.5) +
  MoMAColors::scale_fill_moma_d('Rattner') +
  geom_text_repel(aes(y = Allele_Frequency, 
                      x = POS, 
                      label = Accessions) ) +
  xlim(7878906, 7885876) + # set the same limits than the other plot 
  ylim(0, 1) +
  theme_bw() +
  xlab("Chromosome 4") +
  ylab("Allele Frequency") + 
  theme(
    legend.text = element_text(size = 10),
    legend.title = element_blank(),
    legend.position = "bottom"
    ) +
  # Add SNP position
  geom_vline(xintercept = 7882960, color = "red", linetype = "dashed", linewidth = .5)
```

```{r}
KSL4_2_plot / SNPs_plot
```

### Plot O.officinalis cluster

```{r}
Oo_MBGC <- 
  O_officinalis %>%
  mutate(Orientation = str_replace(Orientation, "\\+", 'TRUE')) %>%
  mutate(Orientation = str_replace(Orientation, "\\-", 'FALSE')) %>%
  ggplot(
    aes(xmin = Start, 
        xmax = End, 
        y = Genome, 
        fill = Gene_Name, 
        forward = Orientation,
        label = Gene_Name)
  ) +
  geom_gene_arrow(
                  arrowhead_height = unit(5, "mm"), 
                  arrowhead_width = unit(1, "mm"),
                  arrow_body_height = grid::unit(5, "mm"),
                  linetype = 'solid',
                  size = 0.2) +
  facet_wrap(~Genome, 
             scales = "free_y", 
             ncol = 1,
             labeller = label_parsed) +
  scale_fill_manual(values = c("#EE8866", "#44BB99", "#EBEBEB", "#EEDD88", "#99DDFF")) +
  theme_genes() +
  theme(panel.grid.major.y = element_line(colour = NULL)) + 
  theme(
    legend.text = element_text(size = 10),
    legend.title = element_blank(),
    legend.position = "top",
    axis.title.x=element_blank(),
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(), 
    axis.line.x = element_blank(),
    axis.title.y=element_blank(),
    axis.ticks.y=element_blank(), 
    axis.line.y = element_blank() )

```

```{r}
Oo_MBGC / KSL4_2_plot / SNPs_plot
```

```{r}
ggsave('plots/snp_calling.pdf',
      dpi = 400,
      width = 10, 
      height = 5) 
```